### ISSUE TITLE
Idea: transform LogicalPlans into sqlparser AST Query nodes

### ISSUE THREAD

**Issue by jonmmease** [nevi-me/datafusion-rdbms-ext#8]
Rather than introduce plan nodes for each query operation (`SqlJoinPlanNode`, `SqlProjectAggregateNode`, etc.), I was wondering if it would make sense to reuse the AST struct hierarchy defined by `sqlparser`.  Specifically, would it make sense to write a (fallible) transform from a `LogicalPlan` struct to an `sqlparser::ast::Query` struct?

https://github.com/sqlparser-rs/sqlparser-rs/blob/12a3e97ef39333c82afaae3b377c2166db8ec9c4/src/ast/query.rs#L23-L40

The idea would be to use a single `SqlPlanNode` that would hold a `Query` struct.  The optimizer would convert supported branches of the logical plan into `SqlPlanNode`s, then the planner would generate the SQL string from the `Query` struct using the `Display` logic in `sqlparser`.  As part of the optimization logic, candidate `Query` structs could be passed to the current `DatabaseConnector`, allowing the connector to reject unsupported queries

One potential downside of this approach is that the `Query` struct's `Display` logic doesn't consider the particular SQL dialect. But I wonder if the subset of SQL we'd generated from logical plans would even require dialect distinctions.

**Comment by nevi-me** [comment]
> Rather than introduce plan nodes for each query operation (`SqlJoinPlanNode`, `SqlProjectAggregateNode`, etc.), 

I didn't think about the design much, but also noticed that the nodes introduce mental overhead as they're effectively some way of beign able to generate the SQL to pass toe the DB, but also delaying materialising the SQL query (with `to_sql()`) until query time.

> I was wondering if it would make sense to reuse the AST struct hierarchy defined by `sqlparser`. Specifically, would it make sense to write a (fallible) transform from a `LogicalPlan` struct to an `sqlparser::ast::Query` struct?

That sounds like a better implementation than what I've done with the `to_sql()` on `Expr` and `LogicalPlan`.
I haven't used sqlparser before, so I'm unfamiliar with its internals. Would `sqlparser::ast::Query` suffice to also cover joins? I can't immediately see how `sqlparser::ast::Join` relates to `Query`.

I think on the optimisation part, it would still be useful to have some rules (some can be cost-based) that determine which plans can be pushed to the source.

For example,

* `LogicalPlan::Aggregate` will almost always reduce the volume of data copied from source, so all supported calculations should be pushed to source.
* `select col1, col2, calc1 from table` could easily push `calc1` to the source, but we might not want that as the source would compute the value, and we'd have an extra column to transmit.

So maybe we can break things down into a single `SqlPlanNode`, and then work out the optimization rules and improvements separately.

What do you think?

**Comment by jonmmease** [comment]
> Would sqlparser::ast::Query suffice to also cover joins? I can't immediately see how sqlparser::ast::Join relates to Query

Yeah, for Joins: `Query` holds a `SetExpr`. `SetExpr` is an enum with a `Select` variant. The `Select` struct has a `from` field with holds a `Vec` of `TableWithJoins` structs. `TableWithJoins` holds a `Vec` of `Join` struct.

It's probably a bit more nested and verbose that what we would strictly need to represent the features of a DataFusion logical plan.  But my feeling is that the advantage of knowing that it has everything we need, and that we don't need to maintain our own version of an SQL AST, is worth it.

> I think on the optimisation part, it would still be useful to have some rules (some can be cost-based) that determine which plans can be pushed to the source.

Yeah, I think it makes sense for the optimizer(s) to make decisions on what's worth pushing to the external database. I haven't given much thought to the situations where it might be advantageous *not* to push compatible computations to the external database, but your `SELECT` example makes sense.

**Comment by nevi-me** [comment]
I've been slowly working on this @jonmmease, it's a bit tedious so far converting LogicalPlan back to sqlparser::ast, but I see the benefits of the work. There's some datafusion inconveniences around how column names are generated, e.g. sum(col_a - col_b * 3) generates a messy name. In an aggregate select with filters, I'm close to converting that to its original query.

I'll push my changes into a branch one of these days
